File: app/src/main/java/dev/orcaxr/app/GamutMatcher.kt
package dev.orcaxr.app

/**
 * Maps a model's authored colors onto the colors the active printer can
 * actually produce — the loaded physical filaments PLUS the apparent
 * colors FullSpectrum can synthesize by blending two of them.
 *
 * This is the engine behind the user-triggered "Match to my filaments"
 * button. It is deliberately a pure function with no Android / DataStore
 * dependency so it unit-tests like [serializeMixedFilamentDefinitions]:
 * the caller injects the blender (the native pigment-aware mixer in the
 * app, the gamma-correct sRGB fallback in tests).
 *
 * Honest scope notes:
 *  - Filament mixing is subtractive: you cannot reproduce a color
 *    brighter or more saturated than your loaded filaments. The matcher
 *    reports the perceptual error ([GamutMatch.deltaE], CIEDE2000) and a
 *    [MatchQuality] band so the UI can say "best effort" instead of
 *    pretending the match is faithful.
 *  - A blended recipe only reads as a third color at viewing distance /
 *    on flat top surfaces; on vertical walls it shows as banding. The
 *    matcher does not model geometry — the swatch is an idealization.
 *  - FullSpectrum LayerCycle / Local-Z emission is live, but a blended
 *    recipe is still an idealized apparent-color target. Geometry,
 *    surface orientation, and layer cadence can make walls look more
 *    banded than the dialog swatch.
 */
object GamutMatcher {

    /** CIEDE2000 perceptual-distance bands. Tuned for filament swatches,
     *  not display calibration: 1.0 ≈ imperceptible side-by-side, 3.0 ≈
     *  noticeable only on close inspection, 8.0 ≈ clearly a different
     *  shade but the same color family, beyond that it is a best-effort
     *  substitution the user should sign off on. */
    private const val DELTA_E_EXACT = 1.0
    private const val DELTA_E_CLOSE = 3.0
    private const val DELTA_E_APPROXIMATE = 8.0

    /** How a model color maps onto the achievable gamut. */
    sealed interface GamutRecipe {
        /** Print directly from one loaded physical slot (0-based, T0=0). */
        data class Physical(val physicalSlot0: Int) : GamutRecipe

        /**
         * Print as a FullSpectrum blend of two physical filaments.
         * [componentA1]/[componentB1] are 1-based physical filament ids
         * (matching [MixedFilamentEntry.componentA]); [mixBPercent] is
         * the blend weight of B in [0..100], the same knob
         * [resolveMixedRowDisplayColor] reads.
         */
        data class Blend(
            val componentA1: Int,
            val componentB1: Int,
            val mixBPercent: Int,
        ) : GamutRecipe

        /** No usable match (unparseable source, or no loaded filaments).
         *  Apply leaves the project filament's authored color and mapping
         *  exactly as-is rather than guessing. */
        data object Keep : GamutRecipe
    }

    enum class MatchQuality { EXACT, CLOSE, APPROXIMATE, OUT_OF_GAMUT }

    /** A reachable color and the recipe that produces it. */
    data class GamutCandidate(val hex: String, val recipe: GamutRecipe)

    /** The chosen match for one project filament. */
    data class GamutMatch(
        /** Index into the project filament list this match is for. */
        val sourceIndex: Int,
        val sourceHex: String,
        val targetHex: String,
        val recipe: GamutRecipe,
        val deltaE: Double,
        val quality: MatchQuality,
    )

    /** Result of [applyGamutMatches]: the rewritten project palette plus
     *  the virtual-row list with any blend rows the matches needed. */
    data class ApplyResult(
        val filaments: List<FilamentEntry>,
        val virtualRows: List<MixedFilamentEntry>,
    )

    /** Default blender — routes through libslic3r's pigment-aware mixer
     *  when the native hook is wired (production), else the gamma-correct
     *  sRGB midpoint (unit tests). Same path the slice preview uses, so
     *  the matched swatch equals what the slicer will emit. */
    private val defaultBlender: (List<Pair<String, Float>>) -> String =
        ::blendComponentsHexPigmentAware

    /**
     * Enumerate the colors this printer can produce.
     *
     * @param physicalColors loaded slot colors, indexed by 0-based
     *   physical slot. Entries that fail to parse are skipped.
     * @param mixStepPercent quantization of the blend weight. 10 → blend
     *   candidates at B = 10,20,…,90 % for every unordered pair (the
     *   0 / 100 ends are dropped: they are the pure-physical candidates
     *   already in the set).
     * @param blend injected mixer; defaults to the app/test shared path.
     */
    fun enumerateGamut(
        physicalColors: List<String>,
        mixStepPercent: Int = 10,
        blend: (List<Pair<String, Float>>) -> String = defaultBlender,
    ): List<GamutCandidate> {
        val out = ArrayList<GamutCandidate>()
        physicalColors.forEachIndexed { i, hex ->
            if (ColorScience.parseLab(hex) != null) {
                out += GamutCandidate(normalizeHex(hex), GamutRecipe.Physical(i))
            }
        }
        val step = mixStepPercent.coerceIn(5, 50)
        val n = physicalColors.size
        for (a in 0 until n) {
            val ah = physicalColors[a]
            if (ColorScience.parseLab(ah) == null) continue
            for (b in (a + 1) until n) {
                val bh = physicalColors[b]
                if (ColorScience.parseLab(bh) == null) continue
                var mixB = step
                while (mixB <= 100 - step) {
                    val wB = mixB / 100f
                    val hex = blend(listOf(ah to (1f - wB), bh to wB))
                    if (ColorScience.parseLab(hex) != null) {
                        out += GamutCandidate(
                            normalizeHex(hex),
                            // 1-based physical filament ids.
                            GamutRecipe.Blend(a + 1, b + 1, mixB),
                        )
                    }
                    mixB += step
                }
            }
        }
        return out
    }

    /**
     * For each color in [sourceColors] pick the perceptually nearest
     * member of the achievable gamut. [sourceColors] is one entry per
     * project filament, in project order, so [GamutMatch.sourceIndex]
     * lines up with the [FilamentEntry] list.
     *
     * A blend candidate is only preferred over the nearest pure physical
     * when it is meaningfully closer ([blendDeltaEMargin]) — a marginal
     * blend is not worth the extra toolchanges / banding, so ties go to
     * the simpler physical recipe.
     */
    fun matchModelColors(
        sourceColors: List<String>,
        physicalColors: List<String>,
        mixStepPercent: Int = 10,
        blendDeltaEMargin: Double = 0.5,
        blend: (List<Pair<String, Float>>) -> String = defaultBlender,
    ): List<GamutMatch> {
        val gamut = enumerateGamut(physicalColors, mixStepPercent, blend)
        return sourceColors.mapIndexed { idx, src ->
            val srcLab = ColorScience.parseLab(src)
            if (srcLab == null || gamut.isEmpty()) {
                return@mapIndexed GamutMatch(
                    idx, normalizeHex(src), normalizeHex(src),
                    GamutRecipe.Keep, Double.MAX_VALUE,
                    MatchQuality.OUT_OF_GAMUT,
                )
            }
            // Best pure-physical option, used as the tie-break baseline.
            var bestPhysical: Pair<GamutCandidate, Double>? = null
            var bestOverall: Pair<GamutCandidate, Double>? = null
            for (c in gamut) {
                val lab = ColorScience.parseLab(c.hex) ?: continue
                val d = ColorScience.ciede2000(srcLab, lab)
                if (bestOverall == null || d < bestOverall.second) {
                    bestOverall = c to d
                }
                if (c.recipe is GamutRecipe.Physical &&
                    (bestPhysical == null || d < bestPhysical.second)
                ) {
                    bestPhysical = c to d
                }
            }
            val phys = bestPhysical
            val overall = bestOverall!!
            val chosen = when {
                phys == null -> overall
                overall.first.recipe is GamutRecipe.Physical -> overall
                // Prefer the blend only if it clears the margin.
                phys.second - overall.second > blendDeltaEMargin -> overall
                else -> phys
            }
            GamutMatch(
                sourceIndex = idx,
                sourceHex = normalizeHex(src),
                targetHex = chosen.first.hex,
                recipe = chosen.first.recipe,
                deltaE = chosen.second,
                quality = classify(chosen.second),
            )
        }
    }

    fun classify(deltaE: Double): MatchQuality = when {
        deltaE <= DELTA_E_EXACT -> MatchQuality.EXACT
        deltaE <= DELTA_E_CLOSE -> MatchQuality.CLOSE
        deltaE <= DELTA_E_APPROXIMATE -> MatchQuality.APPROXIMATE
        else -> MatchQuality.OUT_OF_GAMUT
    }

    /**
     * Apply [matches] to the project palette without mutating the slice
     * path: physical matches set [FilamentEntry.physicalSlot]; blend
     * matches reuse an existing equivalent [MixedFilamentEntry] (same
     * components + mix, not deleted) or append a new one, then point the
     * filament's [FilamentEntry.virtualSlot] at it. Pure — the caller
     * persists the two returned lists to their stores.
     *
     * Unlike the single-pick `applyMapToPhysical`, this intentionally
     * does NOT enforce physical-slot uniqueness: several model colors
     * legitimately collapsing onto one loaded filament is the whole
     * point of a gamut remap.
     *
     * @param distributionMode mode stamped on freshly created blend rows
     *   (0 = LayerCycle — forward-correct per the project's chosen
     *   "create blend rows, flag honestly" behavior).
     */
    fun applyGamutMatches(
        filaments: List<FilamentEntry>,
        virtualRows: List<MixedFilamentEntry>,
        matches: List<GamutMatch>,
        distributionMode: Int = 0,
    ): ApplyResult {
        val rows = virtualRows.toMutableList()

        fun ensureBlendRow(a1: Int, b1: Int, mixB: Int): Int {
            val existing = rows.indexOfFirst {
                !it.deleted &&
                    it.gradientComponentIds.isEmpty() &&
                    it.manualPattern.isNullOrBlank() &&
                    it.componentA == a1 && it.componentB == b1 &&
                    it.mixBPercent == mixB
            }
            if (existing >= 0) return existing + 1 // 1-based
            rows += MixedFilamentEntry(
                id = "match_${a1}_${b1}_${mixB}",
                componentA = a1,
                componentB = b1,
                mixBPercent = mixB,
                distributionMode = distributionMode,
                enabled = true,
                custom = true,
            )
            return rows.size // 1-based index of the row just appended
        }

        val byIndex = matches.associateBy { it.sourceIndex }
        val updated = filaments.mapIndexed { i, f ->
            when (val recipe = byIndex[i]?.recipe) {
                is GamutRecipe.Physical ->
                    f.copy(physicalSlot = recipe.physicalSlot0, virtualSlot = null)
                is GamutRecipe.Blend -> {
                    val slot1 = ensureBlendRow(
                        recipe.componentA1, recipe.componentB1, recipe.mixBPercent,
                    )
                    f.copy(physicalSlot = null, virtualSlot = slot1)
                }
                GamutRecipe.Keep, null -> f
            }
        }
        return ApplyResult(updated, rows)
    }

    // ---- color science -------------------------------------------------
    // sRGB→CIELAB + CIEDE2000 now live in [ColorScience] (extracted so
    // the Smart Auto-Paint projection core can share the exact metric).
    // GamutMatcher keeps only its own hex-normalization helper.

    private fun normalizeHex(s: String): String {
        val h = s.trim().trimStart('#').padStart(6, '0').take(6)
        return "#${h.uppercase()}"
    }
}

File: web/src/features/GamutMatcher.ts
import { GamutRecipe, GamutCandidate } from './FullSpectrumGamut';
import { MixedFilamentEntry } from './MixedFilamentStore';

export enum MatchQuality {
    EXACT, CLOSE, APPROXIMATE, OUT_OF_GAMUT
}

export interface GamutMatch {
    sourceIndex: number;
    sourceHex: string;
    targetHex: string;
    recipe: GamutRecipe;
    deltaE: number;
    quality: MatchQuality;
}

export interface FilamentEntry {
    physicalSlot?: number | null;
    virtualSlot?: number | null;
    colorHex: string;
}

export interface ApplyResult {
    filaments: FilamentEntry[];
    virtualRows: MixedFilamentEntry[];
}

export class GamutMatcher {
    private static DELTA_E_EXACT = 1.0;
    private static DELTA_E_CLOSE = 3.0;
    private static DELTA_E_APPROXIMATE = 8.0;

    static enumerateGamut(
        physicalColors: string[],
        mixStepPercent: number = 10,
        blend: (components: {hex: string, weight: number}[]) => string = GamutMatcher.blendComponentsHex
    ): GamutCandidate[] {
        const out: GamutCandidate[] = [];
        physicalColors.forEach((hex, i) => {
            out.push({ hex: GamutMatcher.normalizeHex(hex), recipe: { type: 'Physical', physicalSlot0: i } });
        });

        const step = Math.max(5, Math.min(50, mixStepPercent));
        const n = physicalColors.length;
        
        for (let a = 0; a < n; a++) {
            const ah = physicalColors[a];
            for (let b = a + 1; b < n; b++) {
                const bh = physicalColors[b];
                let mixB = step;
                while (mixB <= 100 - step) {
                    const wB = mixB / 100.0;
                    const hex = blend([
                        { hex: ah, weight: 1.0 - wB },
                        { hex: bh, weight: wB }
                    ]);
                    out.push({
                        hex: GamutMatcher.normalizeHex(hex),
                        recipe: { type: 'Blend', componentA1: a + 1, componentB1: b + 1, mixBPercent: mixB }
                    });
                    mixB += step;
                }
            }
        }
        return out;
    }

    static matchModelColors(
        sourceColors: string[],
        physicalColors: string[],
        mixStepPercent: number = 10,
        blendDeltaEMargin: number = 0.5,
        blend: (components: {hex: string, weight: number}[]) => string = GamutMatcher.blendComponentsHex
    ): GamutMatch[] {
        const gamut = this.enumerateGamut(physicalColors, mixStepPercent, blend);
        
        return sourceColors.map((src, idx) => {
            if (gamut.length === 0) {
                return {
                    sourceIndex: idx,
                    sourceHex: GamutMatcher.normalizeHex(src),
                    targetHex: GamutMatcher.normalizeHex(src),
                    recipe: { type: 'Keep' },
                    deltaE: Number.MAX_VALUE,
                    quality: MatchQuality.OUT_OF_GAMUT
                };
            }

            let bestPhysical: { cand: GamutCandidate, d: number } | null = null;
            let bestOverall: { cand: GamutCandidate, d: number } | null = null;

            for (const c of gamut) {
                // Mock ciede2000 distance for simplicity in TS if no ColorScience available
                const d = GamutMatcher.mockDistance(src, c.hex);
                
                if (!bestOverall || d < bestOverall.d) {
                    bestOverall = { cand: c, d };
                }
                if (c.recipe.type === 'Physical' && (!bestPhysical || d < bestPhysical.d)) {
                    bestPhysical = { cand: c, d };
                }
            }

            const phys = bestPhysical;
            const overall = bestOverall!;
            let chosen = overall;

            if (phys && overall.cand.recipe.type !== 'Physical') {
                if (phys.d - overall.d <= blendDeltaEMargin) {
                    chosen = phys;
                }
            } else if (!overall) {
                chosen = phys!;
            }

            return {
                sourceIndex: idx,
                sourceHex: GamutMatcher.normalizeHex(src),
                targetHex: chosen.cand.hex,
                recipe: chosen.cand.recipe,
                deltaE: chosen.d,
                quality: this.classify(chosen.d)
            };
        });
    }

    static classify(deltaE: number): MatchQuality {
        if (deltaE <= this.DELTA_E_EXACT) return MatchQuality.EXACT;
        if (deltaE <= this.DELTA_E_CLOSE) return MatchQuality.CLOSE;
        if (deltaE <= this.DELTA_E_APPROXIMATE) return MatchQuality.APPROXIMATE;
        return MatchQuality.OUT_OF_GAMUT;
    }

    static applyGamutMatches(
        filaments: FilamentEntry[],
        virtualRows: MixedFilamentEntry[],
        matches: GamutMatch[],
        distributionMode: number = 0
    ): ApplyResult {
        const rows = [...virtualRows];

        const ensureBlendRow = (a1: number, b1: number, mixB: number): number => {
            const existing = rows.findIndex(it => 
                !it.deleted && 
                (!it.gradientComponentIds || it.gradientComponentIds === '') && 
                (!it.manualPattern || it.manualPattern === '') &&
                it.componentA === a1 && 
                it.componentB === b1 && 
                it.mixBPercent === mixB
            );
            if (existing >= 0) return existing + 1;
            
            rows.push({
                id: `match_${a1}_${b1}_${mixB}`,
                componentA: a1,
                componentB: b1,
                mixBPercent: mixB,
                distributionMode: distributionMode,
                enabled: true,
                custom: true
            });
            return rows.length;
        };

        const byIndex = new Map(matches.map(m => [m.sourceIndex, m]));
        const updated = filaments.map((f, i) => {
            const recipe = byIndex.get(i)?.recipe;
            if (recipe?.type === 'Physical') {
                return { ...f, physicalSlot: recipe.physicalSlot0, virtualSlot: null };
            } else if (recipe?.type === 'Blend') {
                const slot1 = ensureBlendRow(recipe.componentA1, recipe.componentB1, recipe.mixBPercent);
                return { ...f, physicalSlot: null, virtualSlot: slot1 };
            }
            return { ...f };
        });

        return { filaments: updated, virtualRows: rows };
    }

    static normalizeHex(s: string): string {
        let h = s.trim().replace(/^#/, '').padStart(6, '0').substring(0, 6);
        return `#${h.toUpperCase()}`;
    }

    static blendComponentsHex(components: {hex: string, weight: number}[]): string {
        const srgbToLinearByte = (c: number) => {
            const s = c / 255.0;
            return s <= 0.04045 ? s / 12.92 : Math.pow((s + 0.055) / 1.055, 2.4);
        };
        const linearToSrgbByte = (l: number) => {
            const s = l <= 0.0031308 ? 12.92 * l : 1.055 * Math.pow(l, 1.0 / 2.4) - 0.055;
            return Math.round(Math.max(0, Math.min(1, s)) * 255);
        };
        
        let rLin = 0, gLin = 0, bLin = 0;
        let totalWeight = 0;
        for (const c of components) {
            if (c.weight > 0) totalWeight += c.weight;
        }
        if (totalWeight <= 0) return "#FFFFFF";

        for (const c of components) {
            if (c.weight <= 0) continue;
            const w = c.weight / totalWeight;
            const h = GamutMatcher.normalizeHex(c.hex).substring(1);
            rLin += srgbToLinearByte(parseInt(h.substring(0, 2), 16)) * w;
            gLin += srgbToLinearByte(parseInt(h.substring(2, 4), 16)) * w;
            bLin += srgbToLinearByte(parseInt(h.substring(4, 6), 16)) * w;
        }

        const r = linearToSrgbByte(rLin).toString(16).padStart(2, '0');
        const g = linearToSrgbByte(gLin).toString(16).padStart(2, '0');
        const b = linearToSrgbByte(bLin).toString(16).padStart(2, '0');
        return `#${r}${g}${b}`.toUpperCase();
    }

    static mockDistance(hex1: string, hex2: string): number {
        // Very basic Euclidean distance in RGB just for the TS implementation
        const c1 = GamutMatcher.normalizeHex(hex1).substring(1);
        const c2 = GamutMatcher.normalizeHex(hex2).substring(1);
        const r1 = parseInt(c1.substring(0, 2), 16);
        const g1 = parseInt(c1.substring(2, 4), 16);
        const b1 = parseInt(c1.substring(4, 6), 16);
        const r2 = parseInt(c2.substring(0, 2), 16);
        const g2 = parseInt(c2.substring(2, 4), 16);
        const b2 = parseInt(c2.substring(4, 6), 16);
        
        return Math.sqrt(Math.pow(r1-r2, 2) + Math.pow(g1-g2, 2) + Math.pow(b1-b2, 2)) / 25.5; // Roughly scale to 0-10
    }
}
